### PR TITLE
Add custom image to create server group

### PR DESCRIPTION
**Pull request by JargoonPard**
Added the ability to specify a custom user image in a storage account to use for the server group.

@spinnaker/netflix-reviewers @spinnaker/google-reviewers PTAL

### CODE REVIEW

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/vmimage/model/AzureNamedImage.groovy [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package com.netflix.spinnaker.clouddriver.azure.resources.vmimage.model
+
+/**
```
Comment (duftler) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/vmimage/model/AzureNamedImage.groovy]: We're typically leaving out author and related javadoc comments (guessing your ide dropped this here).

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/vmimage/model/AzureNamedImage.groovy [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+ * Created by larry on 3/10/16.
+ */
+class AzureNamedImage {
+  String imageName
+  Boolean iscustom = false
```
Comment (duftler) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/vmimage/model/AzureNamedImage.groovy]: I know you're just moving this class here, but it reads a bit funny to have `iscustom` as opposed to `isCustom`.

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/templates/AzureServerGroupResourceTemplate.groovy [hunk 3]
```diff
@@ -52,18 +54,27 @@ class AzureServerGroupResourceTemplate {
      */
     ServerGroupTemplate(AzureServerGroupDescription description) {
       parameters = new ServerGroupTemplateParameters()
-      variables = new ServerGroupTemplateVariables(description)
 
-      resources.add(new StorageAccount(description))
+      if (description.image.iscustom) {
```
Comment (duftler) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/templates/AzureServerGroupResourceTemplate.groovy]: Can `image` ever be null?

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/templates/AzureServerGroupResourceTemplate.groovy [hunk 4]
```diff
@@ -277,7 +285,8 @@ class AzureServerGroupResourceTemplate {
      * @param description
      */
     ScaleSetOsProfileProperty(AzureServerGroupDescription description) {
-      computerNamePrefix = description.getIdentifier()
+      //Max length of 10 characters to allow for an aditional postfix within a max length of 15 characters
+      computerNamePrefix = description.getIdentifier().substring(0,10)
```
Comment (duftler) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/templates/AzureServerGroupResourceTemplate.groovy]: Please add a space after the `,`.

Comment (duftler) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/templates/AzureServerGroupResourceTemplate.groovy]: Would it make sense to log the truncation here?

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/templates/AzureServerGroupResourceTemplate.groovy [hunk 5]
```diff
@@ -390,23 +399,37 @@ class AzureServerGroupResourceTemplate {
...
    */
   static class ScaleSetVMProfileProperty {
-    ScaleSetStorageProfile storageProfile
+    StorageProfile storageProfile
     ScaleSetOsProfileProperty osProfile
     ScaleSetNetworkProfileProperty networkProfile
 
     ScaleSetVMProfileProperty(AzureServerGroupDescription description) {
-      storageProfile = new ScaleSetStorageProfile(description)
+      if (description.image.iscustom) {
```
Comment (duftler) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/templates/AzureServerGroupResourceTemplate.groovy]: Ternary operator might read neater here.

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/templates/AzureServerGroupResourceTemplate.groovy [hunk 6]
```diff
@@ -437,4 +476,21 @@ class AzureServerGroupResourceTemplate {
...
+    String createOption
+    String osType
+    Map<String, String> image = [:]
+
+    VirtualMachineCustomImageOSDisk (AzureServerGroupDescription description) {
+      name = "osdisk-" + description.getIdentifier()
+      caching = "ReadOnly"
+      createOption = "FromImage"
+      osType = description.image.ostype
+      image["uri"] = description.image.uri
```
Comment (duftler) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/templates/AzureServerGroupResourceTemplate.groovy]: Should be able to use `image.uri = `.

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/templates/AzureServerGroupResourceTemplate.groovy [hunk 7]
```diff
@@ -437,4 +476,21 @@ class AzureServerGroupResourceTemplate {
...
+  static class VirtualMachineCustomImageOSDisk implements OSDisk {
+
+    String name
+    String caching
+    String createOption
+    String osType
+    Map<String, String> image = [:]
+
+    VirtualMachineCustomImageOSDisk (AzureServerGroupDescription description) {
+      name = "osdisk-" + description.getIdentifier()
```
Comment (duftler) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/templates/AzureServerGroupResourceTemplate.groovy]: `name = "osdisk=${description.getIdentifier()}"`

**Comment by JargoonPard** [comment]
Comments addressed and rebased. Thanks for the review @duftler

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/templates/AzureServerGroupResourceTemplate.groovy [hunk 8]
```diff
@@ -277,7 +287,9 @@ class AzureServerGroupResourceTemplate {
      * @param description
      */
     ScaleSetOsProfileProperty(AzureServerGroupDescription description) {
-      computerNamePrefix = description.getIdentifier()
+      //Max length of 10 characters to allow for an aditional postfix within a max length of 15 characters
+      computerNamePrefix = description.getIdentifier().substring(0, 10)
+      log.info("computerNamePrefix will be truncated to 10 chracters to maintain Azure restrictions")
```
Comment (duftler) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/templates/AzureServerGroupResourceTemplate.groovy]: s/chracters/characters

**Comment by duftler** [comment]
Looks like somebody beat us to the commit button, so will need to be
rebased again.

**Comment by JargoonPard** [comment]
rebased. Go! Travis! Go! :-)

**Comment by duftler** [comment]
heh

**Pull request merged**